src/collision/collision_system.hpp

```diff
@@ -56,6 +56,20 @@ class CollisionSystem final
   bool is_free_of_tiles(const Rectf& rect, const bool ignoreUnisolid = false, uint32_t tiletype = Tile::SOLID) const;
   bool is_free_of_statics(const Rectf& rect, const CollisionObject* ignore_object, const bool ignoreUnisolid) const;
   bool is_free_of_movingstatics(const Rectf& rect, const CollisionObject* ignore_object) const;
+
+  struct RaycastResult
+  {
+    const bool is_valid; /**< true if raycast hit something */
+    const bool is_tile = true; /**< true if raycast hit a tile */
+    const CollisionObject* object = nullptr; /**< object that the raycast hit, nullptr if is_tile is true */
+    const Tile* tile = nullptr; /**< tile that the raycast hit, nullptr if is_tile is false */
+    const Rectf tile_box = {}; /**< hitbox of tile, empty if is_tile is false */
+  };
```

Comment:
Move the struct separately from the functions. For example, in a `public:` field right in the beginning of the class.